hooks/hookexecution/execution.go

```diff
@@ -0,0 +1,233 @@
+package hookexecution
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/prebid/prebid-server/hooks"
+	"github.com/prebid/prebid-server/hooks/hookstage"
+)
+
+type hookResponse[T any] struct {
+	Err           error
+	ExecutionTime time.Duration
+	HookID        HookID
+	Result        hookstage.HookResult[T]
+}
+
+type hookHandler[H any, P any] func(
+	context.Context,
+	hookstage.ModuleInvocationContext,
+	H,
+	P,
+) (hookstage.HookResult[P], error)
+
+func executeStage[H any, P any](
+	executionCtx executionContext,
+	plan hooks.Plan[H],
+	payload P,
+	hookHandler hookHandler[H, P],
+) (StageOutcome, P, stageModuleContext, *RejectError) {
+	stageOutcome := StageOutcome{}
+	stageOutcome.Groups = make([]GroupOutcome, 0, len(plan))
+	stageModuleCtx := stageModuleContext{}
+	stageModuleCtx.groupCtx = make([]groupModuleContext, 0, len(plan))
+
+	for _, group := range plan {
+		groupOutcome, newPayload, moduleContexts, reject := executeGroup(executionCtx, group, payload, hookHandler)
+		stageOutcome.ExecutionTimeMillis += groupOutcome.ExecutionTimeMillis
+		stageOutcome.Groups = append(stageOutcome.Groups, groupOutcome)
+		stageModuleCtx.groupCtx = append(stageModuleCtx.groupCtx, moduleContexts)
+		if reject != nil {
+			return stageOutcome, payload, stageModuleCtx, reject
+		}
+
+		payload = newPayload
+	}
+
+	return stageOutcome, payload, stageModuleCtx, nil
+}
+
+func executeGroup[H any, P any](
+	executionCtx executionContext,
+	group hooks.Group[H],
+	payload P,
+	hookHandler hookHandler[H, P],
+) (GroupOutcome, P, groupModuleContext, *RejectError) {
+	var wg sync.WaitGroup
+	done := make(chan struct{})
```

Comment:
Better to rename `done` to its true purpose of `rejected`. It is not a way for `executeHook` to signal work has completed, but rather for `collectHookResponses` to exit pending `executeHook` calls early when a rejection is first encountered.